src/mcm_chapter_10/mcm_ch10_answers.markdown

```diff
@@ -0,0 +1,70 @@
+# MCM Chapter10 Answers
+
+**1. Why does Java define both byte and character streams?**
+
+Java defines both in order to support reading of character based streams, and binary streams. Character streams are unicode based and can be internationalised, and allow the developer convenience in reading in text based files. Byte streams can still be used for reading in characters, in fact character streams are based on byte streams, but offer a number of convenience features, and can in some cases be more efficient to use for text than byte streams.
+
+**2. Even though console input and output is text-based, why does Java still use byte streams for
+this purpose?**
+
+Because the first version of Java only had Byte streams and the System.in() and System.out() is byte stream based, although character streams can also be used.
+
+**3. Show how to open a file for reading bytes.**
+
+See [OpenByteFileDemo.java](io/OpenByteFileDemo.java)
+
+**4. Show how to open a file for reading characters.**
+
+See [OpenCharacterFileDemo.java](io/OpenCharacterFileDemo.java)
+
+**5. Show how to open a file for random-access I/O.**
+
+Use the RandomAccessFile Object to open the file and the seek() operator to go to a location
+
+
+```java
+ 	RandomAccessFile raf = new RandomAccessFile("random.dat", "rw")
+ 
+ 	raf.seek(0);
+	d = raf.readDouble(); //etc...
+```
+
+**6. How can you convert a numeric string such as "123.23" into its binary equivalent?**
+
+There are a number of ways of converting a string number into binary (e.g., see [ConvertStringToBinary.java](io/ConvertStringToBinary.java), however easier ways are to use the Integer Parsing methods e.g.,
```

Comment:
`ConvertStringToBinary.java` appears to convert a string into a binary _string_, rather than into some primitive data format. An example of doing this by hand (inefficiently in this case) might be:
```java
    int parseInt(String num) {
      int out = 0;
      for (int i = num.length()-1; i >= 0; --i) {
        char dec = num.charAt(i);
        if ( i == 0 && dec == '-' ) {
          out = -out;
          break;
        }
        if ( dec < '0' || dec > '9' ) throw new NumberFormatException("Non-integer character found");
        out += (dec - '0') * (Math.pow(10, num.length() - 1 - i));
      }

      return out;
    }
```